### ISSUE TITLE
Odd trailing comma in function call with arrow

### ISSUE THREAD

**Issue by bakkot** [prettier/prettier#2856]
**Prettier 1.7.0**
[Playground link](https://prettier.io/playground/#%7B%22content%22%3A%22%5B%5D.forEach(x%20%3D%3E%20foo(x%2C%20someParticularlyLongParameterNameToMakeThisLineOverflow))%22%2C%22options%22%3A%7B%22ast%22%3Afalse%2C%22bracketSpacing%22%3Atrue%2C%22doc%22%3Afalse%2C%22jsxBracketSameLine%22%3Afalse%2C%22output2%22%3Afalse%2C%22parser%22%3A%22babylon%22%2C%22printWidth%22%3A80%2C%22semi%22%3Atrue%2C%22singleQuote%22%3Afalse%2C%22tabWidth%22%3A2%2C%22trailingComma%22%3A%22all%22%2C%22useTabs%22%3Afalse%7D%7D)
```sh
--trailing-comma all
```

**Input:**
```jsx
[].forEach(x => foo(x, someParticularlyLongParameterNameToMakeThisLineOverflow))
```

**Output:**
```jsx
[].forEach(x =>
  foo(x, someParticularlyLongParameterNameToMakeThisLineOverflow),
);

```

**Expected behavior:**

I think the comma after the `foo()` is weird. It would make some sense if the `x =>` was on the same line as the `foo()`, but since it isn't, it's difficult to read.

**Comment by bakkot** [comment]
Some rational: the main reasons I've heard people advocate trailing commas are that
- they make it easier to swap the order of items in a list, and
- they make for cleaner diffs when appending a new item to the list.

Neither applies in this situation: 
- this case only arises when any other items are all on the same line as the head of the arrow, and
- adding a new item after the arrow will cause the formatting of the arrow to change anyway.

**Comment by suchipi** [comment]
Yeah, this is weird. I actually thought this was a syntax error at first (until I realized it was a trailing comma on the argument to forEach, not a comma after the expression inside the arrow function).

I think we should not print a comma in this case (that is, the function only has one argument, which is an arrow function expression, and the body of the arrow function expression is not a block statement), even with `--trailing-comma all`.

**Comment by vjeux** [comment]
Could you take a look at what eslint does when you enable their flag to enforce all trailing commas?

**Comment by vjeux** [comment]
Also, be aware that it is going to cause some massive number of changes to codebases already using prettier.

**Comment by suchipi** [comment]
With the eslint rule `comma-dangle` set to `["error", "always"]`, eslint doesn't seem to care if there's a comma there; it print no errors with it there, and prints no errors without it there. If you set it to `["error", { "functions": "always" }]`, then it requires it, but it also changes it to this when run with `--fix`:
```js
[].forEach(x, =>
  foo(x, someParticularlyLongParameterNameToMakeThisLineOverflow,),
);
```
which is actually a syntax error. So I'd be surprised if people are using `["error", { "functions": "always" }]` (at least without `arrow-parens` to protect them from the syntax error).

**Comment by bakkot** [comment]
@suchipi

> I think we should not print a comma in this case (that is, the function only has one argument, which is an arrow function expression, and the body of the arrow function expression is not a block statement),

It also comes up when there is [more than one argument](https://prettier.io/playground/#%7B%22content%22%3A%22%5B%5D.forEach(0%2C%20x%20%3D%3E%20foo(x%2C%20someParticularlyLongParameterNameToMakeThisLineOverflow))%22%2C%22options%22%3A%7B%22printWidth%22%3A80%2C%22tabWidth%22%3A2%2C%22singleQuote%22%3Afalse%2C%22trailingComma%22%3A%22all%22%2C%22bracketSpacing%22%3Atrue%2C%22jsxBracketSameLine%22%3Afalse%2C%22parser%22%3A%22babylon%22%2C%22semi%22%3Atrue%2C%22useTabs%22%3Afalse%2C%22doc%22%3Afalse%2C%22ast%22%3Afalse%7D%7D), sometimes.

**Comment by suchipi** [comment]
If there's more than one argument, I'd want to err on the side of printing commas, because that starts to get into the territory of where trailing commas become useful, as you mentioned:

> * they make it easier to swap the order of items in a list, and
> * they make for cleaner diffs when appending a new item to the list.

in eg. the following contrived example:
```js
const doSomeMathSteps = compose(
  x => x + 2,
  x => x * 3,
  x => x / 6,
  // ...
),
```

**Comment by bakkot** [comment]
@suchipi, agreed, in those cases it should still be present. The only case I'm concerned about is when the arguments list does *not* break, but the last arrow in it does: that comes up when there's just one arrow, but also in some other cases; see [the link](https://prettier.io/playground/#%7B%22content%22%3A%22%5B%5D.forEach(0%2C%20x%20%3D%3E%20foo(x%2C%20someParticularlyLongParameterNameToMakeThisLineOverflow))%22%2C%22options%22%3A%7B%22printWidth%22%3A80%2C%22tabWidth%22%3A2%2C%22singleQuote%22%3Afalse%2C%22trailingComma%22%3A%22all%22%2C%22bracketSpacing%22%3Atrue%2C%22jsxBracketSameLine%22%3Afalse%2C%22parser%22%3A%22babylon%22%2C%22semi%22%3Atrue%2C%22useTabs%22%3Afalse%2C%22doc%22%3Afalse%2C%22ast%22%3Afalse%7D%7D) from my other comment.

(This is also the behavior my PR implements.)

**Comment by bakkot** [comment]
@vjeux 

> Also, be aware that it is going to cause some massive number of changes to codebases already using prettier.

Less than you might think: you have to explicitly pass `--trailing-comma=all` for it to come up, and even in projects which do it's pretty rare. For example, this would cause exactly [one](https://github.com/babel/babylon/blob/38504c68e3f0ef93e607e3c40055af0d0f989292/src/plugins/estree.js#L180) change in babylon, five changes in yarn ([x](https://github.com/yarnpkg/yarn/blob/002d39c9b18d9e6e62deb0fe8fc4dc0f3292ccb1/src/cli/commands/upgrade-interactive.js#L62), [x](https://github.com/yarnpkg/yarn/blob/002d39c9b18d9e6e62deb0fe8fc4dc0f3292ccb1/src/config.js#L649), [x](https://github.com/yarnpkg/yarn/blob/002d39c9b18d9e6e62deb0fe8fc4dc0f3292ccb1/__tests__/registries/npm-registry.js#L221), [x](https://github.com/yarnpkg/yarn/blob/002d39c9b18d9e6e62deb0fe8fc4dc0f3292ccb1/__tests__/registries/npm-registry.js#L224), [x](https://github.com/yarnpkg/yarn/blob/002d39c9b18d9e6e62deb0fe8fc4dc0f3292ccb1/__tests__/commands/add.js#L592)) (three of them in test code), and fourteen in react (which I'm not going to link all of, but they all seem [pretty reasonable](https://github.com/facebook/react/blob/dd00b47e1f3aa0e95047a033786a13a5b0cc064f/src/renderers/dom/stack/client/wrappers/ReactDOMTextarea.js#L67-L68)).

React is, what, in the ballpark of a hundred thousand lines of code? That doesn't seem too bad.

---

Incidentally, I just noticed this is a duplicate of #1490, but I'd like to revisit it anyway.

**Comment by vjeux** [comment]
This appears in almost all the product code with jsx:

```js
items.map(x =>
  <SomeJSX>
    ...
  </SomeJSX>
)
```

**Comment by lydell** [comment]
@vjeux I don’t see any trailing commas at all when `.map`ing JSX?

**Prettier 1.7.0**
[Playground link](https://prettier.io/playground/#%7B%22content%22%3A%22items.map(x%20%3D%3E%5Cn%20%20%3CSomeJSX%3E%5Cn%20%20%20%20...%5Cn%20%20%3C%2FSomeJSX%3E%5Cn)%5Cn%20%20%20%20%20%20%20%20%20%20%5Cnitems.map(x%20%3D%3E%5Cn%20%20%3CSomeJSX%3E%5Cn%20%20%20%20with%20some%20more%20text%20this%20time%20to%20force%20a%20break%5Cn%20%20%3C%2FSomeJSX%3E%5Cn)%22%2C%22options%22%3A%7B%22ast%22%3Afalse%2C%22bracketSpacing%22%3Atrue%2C%22doc%22%3Afalse%2C%22jsxBracketSameLine%22%3Afalse%2C%22output2%22%3Afalse%2C%22parser%22%3A%22babylon%22%2C%22printWidth%22%3A80%2C%22semi%22%3Atrue%2C%22singleQuote%22%3Afalse%2C%22tabWidth%22%3A2%2C%22trailingComma%22%3A%22all%22%2C%22useTabs%22%3Afalse%7D%7D)
```sh
--trailing-comma all
```

**Input:**
```jsx
items.map(x =>
  <SomeJSX>
    ...
  </SomeJSX>
)
          
items.map(x =>
  <SomeJSX>
    with some more text this time to force a break
  </SomeJSX>
)
```

**Output:**
```jsx
items.map(x => <SomeJSX>...</SomeJSX>);

items.map(x => (
  <SomeJSX>with some more text this time to force a break</SomeJSX>
));

```

**Comment by kentor** [comment]
Also discussed here https://github.com/prettier/prettier/issues/2512

**Comment by sompylasar** [comment]
May be related: https://github.com/prettier/prettier/issues/3465
